src/api.cpp

```diff
@@ -2793,8 +2793,12 @@ cl_int CLVK_API_CALL clSetKernelArg(cl_kernel kern, cl_uint arg_index,
         return CL_INVALID_ARG_INDEX;
     }
 
+    // With opaque pointers, clspv is unable to infer the type of an unused
+    // kernel argument so just allow nullptr for its value. It will not have an
+    // affect on the kernel's operation.
     if ((arg_value == nullptr) &&
-        (kernel->arg_kind(arg_index) != kernel_argument_kind::local)) {
+        !((kernel->arg_kind(arg_index) != kernel_argument_kind::local) ||
+          (kernel->arg_kind(arg_index) != kernel_argument_kind::unused))) {
```

Comment:
Unless I'm mis-parsing this, I think the logic is wrong here. Didn't you intend `is_null && !(is_local || is_unused)`? Also, I don't think we can do this with the spec as currently written. Negative tests passing say an invalid (but unused) buffer or sampler would fail.

Reply:
Unfortunately with opaque pointers I'm not sure there's much we can do. We can either error for an unused argument and have a false positive for local memory arguments or not error and have a false negative for buffers, images and samplers. I thought it would be better to have the false negatives since the unused argument has no effect.

Reply:
Yeah, we will need a solution to this (maybe using arg info in clvk, maybe by generating descriptor info from arg info in clspv, maybe changing the spec?) at some point but it should not get in the way of real-world applications so I'd rather we unblock you and get forward progress now. Problem for future us :).